File: src/main/java/com/worldbox/sim/Simulation.java
package com.worldbox.sim;

import com.worldbox.config.Config;
import com.worldbox.util.Rng;
import com.worldbox.world.VoxelWorld;
import com.worldbox.world.WorldGen;
import com.worldbox.world.WorldGrid;

public class Simulation {

  public static GameState createInitialState() {
    return createInitialState(Config.WORLD_SEED + (long) (Math.random() * 1_000_000));
  }

  public static GameState createInitialState(long seed) {
    WorldGrid grid = new WorldGrid();
    WorldGen.generate(grid, seed);

    GameState state = new GameState();
    state.grid = grid;
    state.rng = new Rng(seed + 99);
    state.voxels = new VoxelWorld(grid);
    for (int z = 0; z < grid.rows; z++) {
      for (int x = 0; x < grid.cols; x++) state.voxels.resyncHeight(grid, x, z);
    }

    seedWanderers(state, 30);
    Wildlife.spawn(state, seed);
    return state;
  }

  // The world starts with no nations at all - just people. Nations only
  // come into existence when an isolated wanderer (see Population.java)
  // gives up looking for a settlement to join and founds one.
  private static void seedWanderers(GameState state, int count) {
    WorldGrid grid = state.grid;
    int spawned = 0, attempts = 0;
    while (spawned < count && attempts < 3000) {
      attempts++;
      int x = (int) state.rng.range(0, grid.cols);
      int y = (int) state.rng.range(0, grid.rows);
      if (!grid.inBounds(x, y) || grid.terrain[grid.idx(x, y)] == Config.WATER) continue;
      state.humans.add(Population.createAdult(x + 0.5, y + 0.5, -1, -1));
      spawned++;
    }
  }

  /** How many voxel cells' worth of pending water flow to resolve per
   * tick (see VoxelWorld.tickWaterFlow) - cheap per cell (a handful of
   * neighbor lookups), but a huge dug-out area could otherwise queue
   * tens of thousands of them at once, so this is spread out instead of
   * drained in a single tick. */
  private static final int WATER_FLOW_BUDGET = 800;

  public static void tick(GameState state) {
    state.tick++;
    for (long coarseKey : state.voxels.tickWaterFlow(WATER_FLOW_BUDGET)) {
      int cx = (int) (coarseKey % state.grid.cols);
      int cz = (int) (coarseKey / state.grid.cols);
      state.voxels.resyncHeight(state.grid, cx, cz);
      state.grid.markDirtyIdx(state.grid.idx(cx, cz));
    }
    Population.update(state);
    Wildlife.update(state);
    Settlement.update(state);
    Nation.update(state);
    Economy.update(state);
    Government.update(state);
    Military.update(state);
    Diplomacy.update(state);
    Events.update(state);
    Weather.update(state);
  }
}

File: src/sim/simulation.js
import { WorldGrid } from "../world/grid.js";
import { generateWorld, findLandSpot } from "../world/worldgen.js";
import { updatePopulation } from "./population.js";
import { updateSettlements, recomputeTerritory } from "./settlement.js";
import { updateNations, foundNewNation, randomNationName } from "./nation.js";
import { GlobalMarket, updateEconomy } from "./economy.js";
import { updateMilitary } from "./military.js";
import { DiplomacyManager, updateDiplomacy } from "./diplomacy.js";
import { updateEvents } from "./events.js";
import { makeRng } from "../utils/rng.js";
import { WORLD_SEED } from "../config.js";

export function createInitialState(seed = WORLD_SEED + Math.floor(Math.random() * 1e6)) {
  const grid = new WorldGrid();
  generateWorld(grid, seed);

  const state = {
    grid,
    humans: [],
    settlements: new Map(),
    nations: new Map(),
    armies: new Map(),
    market: new GlobalMarket(),
    diplomacy: new DiplomacyManager(),
    tornadoes: [],
    monster: null,
    tick: 0,
    rng: makeRng(seed + 99),
    selection: null,
    paused: false,
    speed: 1,
  };

  seedStartingNations(state, 6);
  recomputeTerritory(state);
  return state;
}

function seedStartingNations(state, count) {
  const grid = state.grid;
  const spots = [];
  let attempts = 0;
  while (spots.length < count && attempts < 400) {
    attempts++;
    const cx = state.rng.range(grid.cols * 0.15, grid.cols * 0.85);
    const cy = state.rng.range(grid.rows * 0.15, grid.rows * 0.85);
    const spot = findLandSpot(grid, cx, cy, 3, state.rng);
    if (!spot) continue;
    const tooClose = spots.some((s) => Math.hypot(s.x - spot.x, s.y - spot.y) < grid.cols * 0.16);
    if (tooClose) continue;
    spots.push(spot);
  }
  for (const spot of spots) {
    foundNewNation(state, spot.x, spot.y, randomNationName(state.rng));
  }
}

export function simulationTick(state) {
  state.tick++;
  updatePopulation(state);
  updateSettlements(state);
  updateNations(state);
  updateEconomy(state);
  updateMilitary(state);
  updateDiplomacy(state);
  updateEvents(state);
}
